middle_end/flambda2/types/provers.ml

```diff
@@ -583,20 +583,23 @@ let prove_unique_fully_constructed_immutable_heap_block env t :
 let meet_is_flat_float_array_value _env
     (value_head : TG.head_of_kind_value_non_null) : bool generic_proof =
   match value_head with
-  | Array { element_kind = Unknown; _ } -> Unknown
-  | Array { element_kind = Bottom; _ } ->
+  | Array { element_kinds = Unknown; _ } -> Unknown
+  | Array { element_kinds = Bottom; _ } ->
     (* Empty array case. We cannot return Invalid, but any other result is
        correct. We arbitrarily pick [false], as this is what we would get if we
        looked at the tag at runtime. *)
     Proved false
-  | Array { element_kind = Ok element_kind; _ } -> (
-    match K.With_subkind.kind element_kind with
-    | Value -> Proved false
-    | Naked_number Naked_float -> Proved true
-    | Naked_number _ -> Invalid
-    | Region | Rec_info ->
-      Misc.fatal_errorf "Wrong element kind for array: %a" K.With_subkind.print
-        element_kind)
+  | Array { element_kinds = Ok element_kinds; _ } -> (
+    match element_kinds with
+    | [] | _ :: _ :: _ -> Invalid
+    | [element_kind] -> (
+      match K.With_subkind.kind element_kind with
+      | Value -> Proved false
+      | Naked_number Naked_float -> Proved true
```

Comment:
That looks wrong. If you look at the comment in `flambda_primitive.mli`, on the `Unboxed_product` constructor of the `Array_kind.t` type, it says that arrays of unboxed products are never float arrays (in the sense of having `Double_array_tag`). So because here we can't distinguish between a legacy flat float array and an array of unboxed floats represented as unboxed products of size 1, we risk returning a wrong answer.

Reply:
The new code here should also be better.